platform/view/services/kvs/kvs.go

```diff
@@ -128,43 +129,31 @@ func (o *KVS) Put(id string, state interface{}) error {
 		return errors.Wrapf(err, "cannot marshal state with id [%s]", id)
 	}
 
-	tx, err := o.store.NewWriteTransaction()
-	if err != nil {
-		return errors.Wrapf(err, "begin update for id [%s] failed", id)
-	}
-
-	logger.Debugf("store [%d] bytes into key [%s:%s]", len(raw), o.namespace, id)
-	if err := tx.SetState(o.namespace, id, raw); err != nil {
-		if err1 := tx.Discard(); err1 != nil {
-			logger.Debugf("got error %v; discarding caused %v", err, err1)
-		}
-
-		if !errors.HasCause(err, driver.UniqueKeyViolation) {
-			return errors.Wrapf(err, "failed to commit value for id [%s]", id)
+	if err := utils.NewProbabilisticRetryRunner(3, 200, true).RunWithErrors(func() (bool, error) {
+		tx, err := o.store.NewWriteTransaction()
+		if err != nil {
+			return true, errors.Wrapf(err, "begin update for id [%s] failed", id)
 		}
-	} else {
-		if err := tx.Commit(); err != nil {
+		logger.Debugf("store [%d] bytes into key [%s:%s]", len(raw), o.namespace, id)
+		if err := tx.SetState(o.namespace, id, raw); err != nil {
 			if err1 := tx.Discard(); err1 != nil {
 				logger.Debugf("got error %v; discarding caused %v", err, err1)
 			}
-			return errors.Wrapf(err, "committing value for id [%s] failed", id)
+
+			if !errors.HasCause(err, driver.UniqueKeyViolation) {
+				return false, errors.Wrapf(err, "failed to commit value for id [%s]", id)
+			}
```

Comment:
We could do `return true, nil` here and avoid the else nesting.

Reply:
okay